### PR TITLE
making more robust the selection of pairs in rve periodicity utility

### PR DESCRIPTION
**Pull request by RiccardoRossi**
**Description**
this minor PR makes more robust the selection of periodicity condition by adding a fallback to the closest node
Also changes the way a tolerane is computed to give the user control and to make it more robust

Please mark the PR with appropriate tags: 
- StructuralMechanicsApplication

**Changelog**
Two changes are made:
1- the tolerance employed in selecting contions is made user-controllable and more inclusive
2- a fallback to the closest node is introduced so to avoid random failures

### CODE REVIEW

File: applications/StructuralMechanicsApplication/custom_utilities/rve_periodicity_utility.cpp [hunk 1]
```diff
@@ -25,7 +25,8 @@ void RVEPeriodicityUtility::AssignPeriodicity(
     ModelPart& rMasterModelPart,
     ModelPart& rSlaveModelPart,
     const Matrix& rStrainTensor,
-    const Vector& rDirection
+    const Vector& rDirection,
+    const double search_tolerance
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_utilities/rve_periodicity_utility.cpp]:
```suggestion
    const double SearchTolerance
```

File: applications/StructuralMechanicsApplication/custom_utilities/rve_periodicity_utility.cpp [hunk 2]
```diff
@@ -46,7 +46,7 @@ void RVEPeriodicityUtility::AssignPeriodicity(
 
         Condition::Pointer p_host_cond;
         Vector N;
-        array_1d<double, 3> transformed_slave_coordinates = it_node->Coordinates() - rDirection;
+        array_1d<double, 3> transformed_slave_coordinates = it_node->GetInitialPosition() - rDirection;
 
         // Finding the host element for this node
         const bool is_found = bin_based_point_locator.FindPointOnMeshSimplified(transformed_slave_coordinates, N, p_host_cond, max_search_results, search_tolerance);
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_utilities/rve_periodicity_utility.cpp]:
```suggestion
        const bool is_found = bin_based_point_locator.FindPointOnMeshSimplified(transformed_slave_coordinates, N, p_host_cond, max_search_results, SearchTolerance);
```

File: applications/StructuralMechanicsApplication/custom_utilities/rve_periodicity_utility.h [hunk 3]
```diff
@@ -103,7 +103,9 @@ class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) RVEPeriodicityUtility
     void AssignPeriodicity(ModelPart& rMasterModelPart,
                            ModelPart& rSlaveModelPart,
                            const Matrix& rStrainTensor,
-                           const Vector& rDirection);
+                           const Vector& rDirection,
+                           const double search_tolerance=1e-6
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_utilities/rve_periodicity_utility.h]:
```suggestion
                           const double SearchTolerance=1e-6
```

File: applications/StructuralMechanicsApplication/custom_utilities/rve_periodicity_utility.cpp [hunk 4]
```diff
@@ -73,7 +73,41 @@ void RVEPeriodicityUtility::AssignPeriodicity(
...
+            double node_distance = 1e50;
+            IndexType closest_node_id = 0;            
+
+            DataTupletype aux_data;
+
+            auto &T = std::get<2>(aux_data);
+            T = translation;
+
+            //brute force search amongst all nodes to find the closest
+            for(auto& rMasterNode : rMasterModelPart.Nodes())
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_utilities/rve_periodicity_utility.cpp]:
```suggestion
            for(auto& rMasterNode : rMasterModelPart.Nodes()) {
```

File: applications/StructuralMechanicsApplication/custom_utilities/rve_periodicity_utility.cpp [hunk 5]
```diff
@@ -73,7 +73,41 @@ void RVEPeriodicityUtility::AssignPeriodicity(
...
+            {
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_utilities/rve_periodicity_utility.cpp]:
```suggestion
```

File: applications/StructuralMechanicsApplication/custom_utilities/rve_periodicity_utility.cpp [hunk 6]
```diff
@@ -73,7 +73,41 @@ void RVEPeriodicityUtility::AssignPeriodicity(
...
+                array_1d<double,3> distance_vector = rMasterNode.GetInitialPosition() - transformed_slave_coordinates;
+                double d = norm_2(distance_vector);
+                if(d < node_distance)
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_utilities/rve_periodicity_utility.cpp]:
```suggestion
                if(d < node_distance) {
```

File: applications/StructuralMechanicsApplication/custom_utilities/rve_periodicity_utility.cpp [hunk 7]
```diff
@@ -73,7 +73,41 @@ void RVEPeriodicityUtility::AssignPeriodicity(
...
+                {
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_utilities/rve_periodicity_utility.cpp]:
```suggestion
```

File: applications/StructuralMechanicsApplication/python_scripts/rve_analysis.py [hunk 8]
```diff
@@ -1,7 +1,7 @@
 import KratosMultiphysics
 import KratosMultiphysics.StructuralMechanicsApplication
 from KratosMultiphysics.StructuralMechanicsApplication.structural_mechanics_analysis import StructuralMechanicsAnalysis
-
+import math
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/python_scripts/rve_analysis.py:6]:
```suggestion

# Importing other libraries
import math
```

**Review by loumalouomega — DISMISSED** [review]
OK

File: applications/StructuralMechanicsApplication/custom_utilities/rve_periodicity_utility.cpp [hunk 9]
```diff
@@ -25,7 +25,8 @@ void RVEPeriodicityUtility::AssignPeriodicity(
     ModelPart& rMasterModelPart,
     ModelPart& rSlaveModelPart,
     const Matrix& rStrainTensor,
-    const Vector& rDirection
+    const Vector& rDirection,
+    const double SearchTolerance
     )
 {
     KRATOS_ERROR_IF(rMasterModelPart.NumberOfConditions() == 0) << "the master is expected to have conditions and it is empty" << std::endl;
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/custom_utilities/rve_periodicity_utility.cpp]:
```suggestion
    KRATOS_ERROR_IF(rMasterModelPart.NumberOfConditions() == 0) << "The master is expected to have conditions and it is empty" << std::endl;
```

**Review by loumalouomega — APPROVED** [review]

**Review by RiccardoRossi — COMMENTED** [review]
thx @loumalouomega , also for the corrections!!

**Pull request merged**